In [1]:
#imports
import pandas as pd
import numpy as np

#csv -> dataframe
species_df = pd.read_csv('species.csv', dtype = str)
parks_df = pd.read_csv('parks.csv')

species_df.head()
#Note: not sure why there is a 14th data column showing-there are only 13 columns of data-this is confirmed
#in the documentation for the dataset. Will come back and drop that.

Species ID             Park Name Category         Order    Family  \
0  ACAD-1000  Acadia National Park   Mammal  Artiodactyla  Cervidae   
1  ACAD-1001  Acadia National Park   Mammal  Artiodactyla  Cervidae   
2  ACAD-1002  Acadia National Park   Mammal     Carnivora   Canidae   
3  ACAD-1003  Acadia National Park   Mammal     Carnivora   Canidae   
4  ACAD-1004  Acadia National Park   Mammal     Carnivora   Canidae   

          Scientific Name                                       Common Names  \
0             Alces alces                                              Moose   
1  Odocoileus virginianus  Northern White-Tailed Deer, Virginia Deer, Whi...   
2           Canis latrans                             Coyote, Eastern Coyote   
3             Canis lupus        Eastern Timber Wolf, Gray Wolf, Timber Wolf   
4           Vulpes vulpes  Black Fox, Cross Fox, Eastern Red Fox, Fox, Re...   

  Record Status     Occurrence  Nativeness Abundance Seasonality  \
0      Approved        Present      Native      Rare    Resident   
1      Approved        Present      Native  Abundant         NaN   
2      Approved        Present  Not Native    Common         NaN   
3      Approved  Not Confirmed      Native       NaN         NaN   
4      Approved        Present     Unknown    Common     Breeder   

  Conservation Status Unnamed: 13  
0                 NaN         NaN  
1                 NaN         NaN  
2  Species of Concern         NaN  
3          Endangered         NaN  
4                 NaN         NaN

In [2]:
#drop that extra column in species_df
species_df = species_df.iloc[: , :-1]
species_df

Species ID             Park Name        Category          Order  \
0       ACAD-1000  Acadia National Park          Mammal   Artiodactyla   
1       ACAD-1001  Acadia National Park          Mammal   Artiodactyla   
2       ACAD-1002  Acadia National Park          Mammal      Carnivora   
3       ACAD-1003  Acadia National Park          Mammal      Carnivora   
4       ACAD-1004  Acadia National Park          Mammal      Carnivora   
...           ...                   ...             ...            ...   
119243  ZION-2791    Zion National Park  Vascular Plant      Solanales   
119244  ZION-2792    Zion National Park  Vascular Plant        Vitales   
119245  ZION-2793    Zion National Park  Vascular Plant        Vitales   
119246  ZION-2794    Zion National Park  Vascular Plant  Zygophyllales   
119247  ZION-2795    Zion National Park  Vascular Plant  Zygophyllales   

                Family         Scientific Name  \
0             Cervidae             Alces alces   
1             Cervidae  Odocoileus virginianus   
2              Canidae           Canis latrans   
3              Canidae             Canis lupus   
4              Canidae           Vulpes vulpes   
...                ...                     ...   
119243      Solanaceae       Solanum triflorum   
119244        Vitaceae         Vitis arizonica   
119245        Vitaceae          Vitis vinifera   
119246  Zygophyllaceae       Larrea tridentata   
119247  Zygophyllaceae     Tribulus terrestris   

                                             Common Names Record Status  \
0                                                   Moose      Approved   
1       Northern White-Tailed Deer, Virginia Deer, Whi...      Approved   
2                                  Coyote, Eastern Coyote      Approved   
3             Eastern Timber Wolf, Gray Wolf, Timber Wolf      Approved   
4       Black Fox, Cross Fox, Eastern Red Fox, Fox, Re...      Approved   
...                                                   ...           ...   
119243                                Cut-Leaf Nightshade      Approved   
119244                                       Canyon Grape      Approved   
119245                                         Wine Grape      Approved   
119246                                      Creosote Bush      Approved   
119247                                      Puncture Vine      Approved   

           Occurrence  Nativeness Abundance Seasonality Conservation Status  
0             Present      Native      Rare    Resident                 NaN  
1             Present      Native  Abundant         NaN                 NaN  
2             Present  Not Native    Common         NaN  Species of Concern  
3       Not Confirmed      Native       NaN         NaN          Endangered  
4             Present     Unknown    Common     Breeder                 NaN  
...               ...         ...       ...         ...                 ...  
119243        Present      Native  Uncommon         NaN                 NaN  
119244        Present      Native  Uncommon         NaN                 NaN  
119245        Present  Not Native  Uncommon         NaN                 NaN  
119246        Present      Native      Rare         NaN                 NaN  
119247        Present  Not Native  Uncommon         NaN                 NaN  

[119248 rows x 13 columns]

In [3]:
#entries where "Record Status" is a Common Name clearly have errors
#entries with "In Review" record status are missing data in some fields,
#may not be as reliable as "Approved" entries
#BUT they make up a substantial portion of the dataset - let's try keeping them in

species_df = species_df[species_df['Record Status'].isin(["Approved", "In Review"])]
species_df

Species ID             Park Name        Category          Order  \
0       ACAD-1000  Acadia National Park          Mammal   Artiodactyla   
1       ACAD-1001  Acadia National Park          Mammal   Artiodactyla   
2       ACAD-1002  Acadia National Park          Mammal      Carnivora   
3       ACAD-1003  Acadia National Park          Mammal      Carnivora   
4       ACAD-1004  Acadia National Park          Mammal      Carnivora   
...           ...                   ...             ...            ...   
119243  ZION-2791    Zion National Park  Vascular Plant      Solanales   
119244  ZION-2792    Zion National Park  Vascular Plant        Vitales   
119245  ZION-2793    Zion National Park  Vascular Plant        Vitales   
119246  ZION-2794    Zion National Park  Vascular Plant  Zygophyllales   
119247  ZION-2795    Zion National Park  Vascular Plant  Zygophyllales   

                Family         Scientific Name  \
0             Cervidae             Alces alces   
1             Cervidae  Odocoileus virginianus   
2              Canidae           Canis latrans   
3              Canidae             Canis lupus   
4              Canidae           Vulpes vulpes   
...                ...                     ...   
119243      Solanaceae       Solanum triflorum   
119244        Vitaceae         Vitis arizonica   
119245        Vitaceae          Vitis vinifera   
119246  Zygophyllaceae       Larrea tridentata   
119247  Zygophyllaceae     Tribulus terrestris   

                                             Common Names Record Status  \
0                                                   Moose      Approved   
1       Northern White-Tailed Deer, Virginia Deer, Whi...      Approved   
2                                  Coyote, Eastern Coyote      Approved   
3             Eastern Timber Wolf, Gray Wolf, Timber Wolf      Approved   
4       Black Fox, Cross Fox, Eastern Red Fox, Fox, Re...      Approved   
...                                                   ...           ...   
119243                                Cut-Leaf Nightshade      Approved   
119244                                       Canyon Grape      Approved   
119245                                         Wine Grape      Approved   
119246                                      Creosote Bush      Approved   
119247                                      Puncture Vine      Approved   

           Occurrence  Nativeness Abundance Seasonality Conservation Status  
0             Present      Native      Rare    Resident                 NaN  
1             Present      Native  Abundant         NaN                 NaN  
2             Present  Not Native    Common         NaN  Species of Concern  
3       Not Confirmed      Native       NaN         NaN          Endangered  
4             Present     Unknown    Common     Breeder                 NaN  
...               ...         ...       ...         ...                 ...  
119243        Present      Native  Uncommon         NaN                 NaN  
119244        Present      Native  Uncommon         NaN                 NaN  
119245        Present  Not Native  Uncommon         NaN                 NaN  
119246        Present      Native      Rare         NaN                 NaN  
119247        Present  Not Native  Uncommon         NaN                 NaN  

[119191 rows x 13 columns]

In [4]:
#what are the possible labels for the "Conservation Status" column?
species_df["Conservation Status"].value_counts()

Species of Concern     3843
Endangered              374
Under Review            194
Threatened              184
In Recovery              77
Proposed Endangered      24
Proposed Threatened       7
Extinct                   1
Name: Conservation Status, dtype: int64

In [5]:
#8 labels in "Conservation Status"; several have only a few entries - outliers may skew data
#let's focus on five main labels

species_df_con = species_df[species_df['Conservation Status'].isin(["Species of Concern", "Endangered", 
                                                                   "Under Review", "Threatened",
                                                                   "In Recovery"])]
species_df_con

Species ID             Park Name        Category       Order  \
2       ACAD-1002  Acadia National Park          Mammal   Carnivora   
3       ACAD-1003  Acadia National Park          Mammal   Carnivora   
20      ACAD-1020  Acadia National Park          Mammal  Chiroptera   
21      ACAD-1021  Acadia National Park          Mammal  Chiroptera   
24      ACAD-1024  Acadia National Park          Mammal  Chiroptera   
...           ...                   ...             ...         ...   
118628  ZION-2176    Zion National Park  Vascular Plant     Fabales   
118840  ZION-2388    Zion National Park  Vascular Plant    Myrtales   
118855  ZION-2403    Zion National Park  Vascular Plant    Myrtales   
118894  ZION-2442    Zion National Park  Vascular Plant      Poales   
118992  ZION-2540    Zion National Park  Vascular Plant      Poales   

                  Family            Scientific Name  \
2                Canidae              Canis latrans   
3                Canidae                Canis lupus   
20      Vespertilionidae           Eptesicus fuscus   
21      Vespertilionidae  Lasionycteris noctivagans   
24      Vespertilionidae              Myotis keenii   
...                  ...                        ...   
118628          Fabaceae       Pediomelum epipsilum   
118840        Onagraceae         Camissonia bairdii   
118855        Onagraceae        Epilobium nevadense   
118894        Cyperaceae               Carex haysii   
118992           Poaceae      Eragrostis pectinacea   

                                             Common Names Record Status  \
2                                  Coyote, Eastern Coyote      Approved   
3             Eastern Timber Wolf, Gray Wolf, Timber Wolf      Approved   
20                        Big Brown Bat, Common Brown Bat      Approved   
21                                      Silver-Haired Bat      Approved   
24      Keen's Myotis, Northern Long-Eared Bat, Northe...      Approved   
...                                                   ...           ...   
118628                                     Kane Breadroot      Approved   
118840                                 Baird's Camissonia      Approved   
118855                                 Nevada Willow-Herb      Approved   
118894                                        Hays' Sedge      Approved   
118992                                   Tufted Lovegrass      Approved   

                        Occurrence  Nativeness Abundance Seasonality  \
2                          Present  Not Native    Common         NaN   
3                    Not Confirmed      Native       NaN         NaN   
20                         Present      Native  Uncommon         NaN   
21                         Present      Native   Unknown         NaN   
24                         Present      Native    Common         NaN   
...                            ...         ...       ...         ...   
118628  Not Present (False Report)      Native       NaN         NaN   
118840                     Present      Native      Rare         NaN   
118855                     Present      Native      Rare         NaN   
118894                     Present      Native  Uncommon         NaN   
118992                     Present      Native  Uncommon         NaN   

       Conservation Status  
2       Species of Concern  
3               Endangered  
20      Species of Concern  
21      Species of Concern  
24      Species of Concern  
...                    ...  
118628  Species of Concern  
118840  Species of Concern  
118855  Species of Concern  
118894  Species of Concern  
118992  Species of Concern  

[4672 rows x 13 columns]

In [6]:
#this data is primarily text
#so we will need to do some label encoding before we can create and train data models

#label encode Park Name
from sklearn.preprocessing import LabelEncoder
class_labels = LabelEncoder()

def LabelEncode (df):
    for col in df:
        df[col] = class_labels.fit_transform(df[col].values)

In [7]:
#label encode species dataframe
LabelEncode(species_df_con)
species_df_con

<ipython-input-6-80ab2d7ab1d4>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[col] = class_labels.fit_transform(df[col].values)


Species ID  Park Name  Category  Order  Family  Scientific Name  \
2                0          0         7     19      42              180   
3                1          0         7     19      42              181   
20               2          0         7     25     270              383   
21               3          0         7     25     270              567   
24               4          0         7     25     270              668   
...            ...        ...       ...    ...     ...              ...   
118628        4667         55        12     42     101              763   
118840        4668         55        12     64     175              172   
118855        4669         55        12     64     175              379   
118894        4670         55        12     79      80              188   
118992        4671         55        12     79     208              384   

        Common Names  Record Status  Occurrence  Nativeness  Abundance  \
2                291              0           4           1          1   
3                355              0           0           0          6   
20               105              0           4           0          4   
21              1021              0           4           0          5   
24               587              0           4           0          1   
...              ...            ...         ...         ...        ...   
118628           581              0           2           0          6   
118840            70              0           4           0          3   
118855           745              0           4           0          3   
118894           512              0           4           0          4   
118992          1127              0           4           0          4   

        Seasonality  Conservation Status  
2                13                    2  
3                13                    0  
20               13                    2  
21               13                    2  
24               13                    2  
...             ...                  ...  
118628           13                    2  
118840           13                    2  
118855           13                    2  
118894           13                    2  
118992           13                    2  

[4672 rows x 13 columns]

In [8]:
#scientific names should be unique - each entry already has unique identifier in "Species ID"
#common names are useful for humans but may just muddy the waters for machine learning/modeling

species_df_filtered = species_df_con.drop(labels = ["Scientific Name", "Common Names"], axis = 1)

species_df_filtered.head()

Species ID  Park Name  Category  Order  Family  Record Status  Occurrence  \
2            0          0         7     19      42              0           4   
3            1          0         7     19      42              0           0   
20           2          0         7     25     270              0           4   
21           3          0         7     25     270              0           4   
24           4          0         7     25     270              0           4   

    Nativeness  Abundance  Seasonality  Conservation Status  
2            1          1           13                    2  
3            0          6           13                    0  
20           0          4           13                    2  
21           0          5           13                    2  
24           0          1           13                    2

In [9]:
#check data types - see if our label encoding behaved
species_df_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4672 entries, 2 to 118992
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   Species ID           4672 non-null   int64
 1   Park Name            4672 non-null   int64
 2   Category             4672 non-null   int64
 3   Order                4672 non-null   int64
 4   Family               4672 non-null   int64
 5   Record Status        4672 non-null   int64
 6   Occurrence           4672 non-null   int64
 7   Nativeness           4672 non-null   int64
 8   Abundance            4672 non-null   int64
 9   Seasonality          4672 non-null   int64
 10  Conservation Status  4672 non-null   int64
dtypes: int64(11)
memory usage: 438.0 KB


In [10]:
#modeling to predict "Conservation Status"
#logistic regression
#import train_test_split
from sklearn.model_selection import train_test_split

X = species_df_filtered.drop('Conservation Status', axis=1)
y = species_df_filtered['Conservation Status']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [11]:
#scale dataset
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [12]:
#set up Logistic Regression model
from sklearn.linear_model import LogisticRegression

regression_con = LogisticRegression()
regression_con.fit(X_train, y_train)

LogisticRegression()

In [13]:
#check performance of regression model
regression_con.fit(X_train, y_train)
y_pred = regression_con.predict(X_test)
regression_con.score(X_test, y_test)

0.8224598930481284

In [14]:
#import MSE - check RMSE
from sklearn.metrics import mean_squared_error as MSE

rmse_log_con = MSE(y_test, y_pred) ** (1/2)
print(rmse_log_con)

0.7370993881569774


In [15]:
#Decision Tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

In [16]:
X = species_df_filtered.drop('Conservation Status', axis=1)
y = species_df_filtered['Conservation Status']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

#scale dataset
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [17]:
dt_con = DecisionTreeClassifier(max_depth=5, random_state=42)
dt_con.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=5, random_state=42)

In [18]:
y_pred = dt_con.predict(X_test)

dt_con_score = accuracy_score(y_test, y_pred)
print(dt_con_score)

0.8245989304812834


In [19]:
#check rmse
rmse_dt_con = MSE(y_test, y_pred) ** (1/2)
print(rmse_dt_con)

0.7718302437283019


In [20]:
#Random Forest
from sklearn.ensemble import RandomForestRegressor

#set seed
SEED = 42

X = species_df_filtered.drop('Conservation Status', axis=1)
y = species_df_filtered['Conservation Status']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = SEED)

#scale dataset
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

In [21]:
rf_con = RandomForestRegressor(n_estimators = 400, min_samples_leaf = 0.1, random_state = SEED)
rf_con.fit(X_train, y_train)

RandomForestRegressor(min_samples_leaf=0.1, n_estimators=400, random_state=42)

In [22]:
y_pred = rf_con.predict(X_test)

rmse_rf_con = MSE(y_test, y_pred) ** (1/2)
print(rmse_rf_con)

0.772536132868872


In [23]:
#neural network imports
import tensorflow
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

X = species_df_filtered.drop('Conservation Status', axis=1)
y = species_df_filtered['Conservation Status']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 42)

#scale dataset
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

number_features = X_train.shape[1]

In [24]:
#defining the model
model_con = Sequential()
model_con.add(Dense(100, activation='relu', kernel_initializer='he_normal', input_shape=(number_features,)))
model_con.add(Dense(100, activation='relu', kernel_initializer='he_normal'))
model_con.add(Dense(6, activation='softmax'))

In [25]:
#compile model
model_con.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [26]:
#run Conservation Status neural network model
model_con.fit(X_train, y_train, epochs=250, batch_size=100, verbose=2)

Epoch 1/250
38/38 - 1s - loss: 1.0905 - accuracy: 0.6944
Epoch 2/250
38/38 - 0s - loss: 0.6628 - accuracy: 0.8234
Epoch 3/250
38/38 - 0s - loss: 0.6166 - accuracy: 0.8255
Epoch 4/250
38/38 - 0s - loss: 0.5966 - accuracy: 0.8271
Epoch 5/250
38/38 - 0s - loss: 0.5886 - accuracy: 0.8271
Epoch 6/250
38/38 - 0s - loss: 0.5760 - accuracy: 0.8269
Epoch 7/250
38/38 - 0s - loss: 0.5670 - accuracy: 0.8285
Epoch 8/250
38/38 - 0s - loss: 0.5578 - accuracy: 0.8309
Epoch 9/250
38/38 - 0s - loss: 0.5507 - accuracy: 0.8298
Epoch 10/250
38/38 - 0s - loss: 0.5498 - accuracy: 0.8282
Epoch 11/250
38/38 - 0s - loss: 0.5434 - accuracy: 0.8285
Epoch 12/250
38/38 - 0s - loss: 0.5336 - accuracy: 0.8328
Epoch 13/250
38/38 - 0s - loss: 0.5294 - accuracy: 0.8303
Epoch 14/250
38/38 - 0s - loss: 0.5250 - accuracy: 0.8314
Epoch 15/250
38/38 - 0s - loss: 0.5194 - accuracy: 0.8341
Epoch 16/250
38/38 - 0s - loss: 0.5140 - accuracy: 0.8346
Epoch 17/250
38/38 - 0s - loss: 0.5124 - accuracy: 0.8354
Epoch 18/250
38/38 - 0s

Epoch 142/250
38/38 - 0s - loss: 0.2765 - accuracy: 0.8970
Epoch 143/250
38/38 - 0s - loss: 0.2801 - accuracy: 0.8999
Epoch 144/250
38/38 - 0s - loss: 0.2798 - accuracy: 0.8954
Epoch 145/250
38/38 - 0s - loss: 0.2805 - accuracy: 0.8970
Epoch 146/250
38/38 - 0s - loss: 0.2780 - accuracy: 0.8999
Epoch 147/250
38/38 - 0s - loss: 0.2736 - accuracy: 0.8980
Epoch 148/250
38/38 - 0s - loss: 0.2798 - accuracy: 0.8967
Epoch 149/250
38/38 - 0s - loss: 0.2717 - accuracy: 0.9021
Epoch 150/250
38/38 - 0s - loss: 0.2721 - accuracy: 0.8986
Epoch 151/250
38/38 - 0s - loss: 0.2685 - accuracy: 0.9010
Epoch 152/250
38/38 - 0s - loss: 0.2709 - accuracy: 0.8999
Epoch 153/250
38/38 - 0s - loss: 0.2645 - accuracy: 0.9069
Epoch 154/250
38/38 - 0s - loss: 0.2722 - accuracy: 0.8975
Epoch 155/250
38/38 - 0s - loss: 0.2705 - accuracy: 0.9026
Epoch 156/250
38/38 - 0s - loss: 0.2679 - accuracy: 0.9010
Epoch 157/250
38/38 - 0s - loss: 0.2677 - accuracy: 0.9029
Epoch 158/250
38/38 - 0s - loss: 0.2662 - accuracy: 0.90

Models worked surprisingly well for Conservation Status - about 92% accuracy with the keras neural network!  Unfortunately, only a small portion of the original data had a Conservation Status value, meaning these models were run on a mere 4672 entries (about 4% of the data used to run the Nativeness models).  In an ideal world, we'd obtain a more complete dataset - with a majority of entries having Conservation Status values - to check these models and/or develop models that better reflect the actual Conservation Statuses of national park flora and fauna.